Service/PublishingService.php

```diff
@@ -0,0 +1,52 @@
+<?php
+declare(strict_types=1);
+
+namespace MageOS\CommonAsyncEvents\Service;
+
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\MessageQueue\PublisherInterface;
+use Magento\Framework\Serialize\Serializer\Json;
+use MageOS\AsyncEvents\Api\AsyncEventRepositoryInterface;
+use MageOS\AsyncEvents\Api\Data\AsyncEventInterface;
+use MageOS\AsyncEvents\Helper\QueueMetadataInterface;
+
+class PublishingService
+{
+    public function __construct(
+        private readonly Json $json,
+        private readonly PublisherInterface $publisher,
+        private readonly AsyncEventRepositoryInterface $asyncEventRepository,
+        private readonly SearchCriteriaBuilder $searchCriteriaBuilder
+    ) {
+    }
+
+    public function publish(string $eventName, array $arguments): void
+    {
+        if (!$this->canSendEvent($eventName)) {
+            return;
+        }
+        $this->publisher->publish(
+            QueueMetadataInterface::EVENT_QUEUE,
+            [
+                $eventName,
+                $this->json->serialize($arguments)
+            ]
+        );
+    }
+
+    private function canSendEvent(string $eventName): bool
+    {
+        $configuredEvents = $this->asyncEventRepository->getList(
```

Comment:
This checks whether a subscriber exists for this event, otherwise the event isn't published to the queue. I will extend it with a caching mechanism later so we save the database call on each entity save, but it shouldn't hurt performance too much.

Reply:
Thanks, if this makes a difference, we should look at moving this into the core publisher.